libavcodec/libproresdecoder.c

```diff
@@ -0,0 +1,237 @@
+    // For now the downscale mode is universally 1x1
+    proresdecodecctx->downscale_mode = kPRDownscale_1x1;
+    
+    // 4 letter tag signal the ProRes profiles. 
+    switch (avctx->codec_tag) { 
+    case MKTAG('a','p','c','o'):
+        avctx->profile = FF_PROFILE_PRORES_PROXY;
+        avctx->bits_per_raw_sample = 10;
+        avctx->pix_fmt = AV_PIX_FMT_YUV422P10LE;
+        proresdecodecctx->pix_fmt = kPRFormat_v210;
+        break;
+    case MKTAG('a','p','c','s'):
+        avctx->profile = FF_PROFILE_PRORES_LT;
+        avctx->bits_per_raw_sample = 10;
+        avctx->pix_fmt = AV_PIX_FMT_YUV422P10LE;
+        proresdecodecctx->pix_fmt = kPRFormat_v210;
+        break;
+    case MKTAG('a','p','c','n'):
+        avctx->profile = FF_PROFILE_PRORES_STANDARD;
+        avctx->bits_per_raw_sample = 10;
+        avctx->pix_fmt = AV_PIX_FMT_YUV422P10LE;
+        proresdecodecctx->pix_fmt = kPRFormat_v210;
+        break;
+    case MKTAG('a','p','c','h'):
+        avctx->profile = FF_PROFILE_PRORES_HQ;
+        avctx->bits_per_raw_sample = 10;
+        avctx->pix_fmt = AV_PIX_FMT_YUV422P10LE;
+        proresdecodecctx->pix_fmt = kPRFormat_v210;
+        break;
+    case MKTAG('a','p','4','h'):
+        avctx->profile = FF_PROFILE_PRORES_4444;
+        avctx->bits_per_raw_sample = 12;
+        avctx->pix_fmt = AV_PIX_FMT_YUVA444P16LE;
+        proresdecodecctx->pix_fmt = kPRFormat_y416;
+        break;
+    case MKTAG('a','p','4','x'):
+        avctx->profile = FF_PROFILE_PRORES_XQ;
+        avctx->bits_per_raw_sample = 12;
+        avctx->pix_fmt = AV_PIX_FMT_YUVA444P16LE;
+        proresdecodecctx->pix_fmt = kPRFormat_y416;
+        break;
+    default:
+        avctx->profile = FF_PROFILE_UNKNOWN;
+        av_log(avctx, AV_LOG_WARNING, "Unknown prores profile %d\n", avctx->codec_tag);
+    }
+
+    if (avctx->bits_per_raw_sample == 10) {
+        av_log(avctx, AV_LOG_DEBUG, "Auto bitdepth precision. Use 10b decoding based on codec tag.\n");
+    } else { /* 12b */
+        av_log(avctx, AV_LOG_DEBUG, "Auto bitdepth precision. Use 12b decoding based on codec tag.\n");
+    }
+
+    proresdecodecctx->m_decoderContext = PROpenDecoder(1, NULL);
+    return 0;
+};
+
+int decodeAppleProResFrame(AVCodecContext *avctx, AVFrame *frame, int *got_frame, AVPacket *avpkt) {
+    ff_thread_finish_setup(avctx);
+    *got_frame = 1; // Should this be set only for +ve return value?
+    return decodePacket(avctx->priv_data, avpkt, frame, 0);
+};
+
+int decodePacket(ProResDecodeCodec* decoderctx, AVPacket* pkt, AVFrame* framePtr, bool drain) {
+    if (drain) {
+        // return AVError_create(AVErrorCode_kErrorEOF);
+        return -1; // error code to be updated
+    } else if (!pkt) {
+        // return AVError_create(AVErrorCode_kErrorEAGAIN);
+        return -2; // error code to be updated
+    }
+
+    int rowBytes = PRBytesPerRowNeededInPixelBuffer(decoderctx->m_width, decoderctx->pix_fmt, decoderctx->downscale_mode);
```

Comment:
Good to have this as we don't have to do this Maths. Although, with the formats explained by Apple in the link above, we could have done that.